lisa/sut_orchestrator/qemu/platform.py

```diff
@@ -423,16 +471,26 @@ def _fill_nodes_metadata(
                 environment, log, qemu_conn, node, timeout
             )
 
+            node_addr = address
+            node_port = 22
+            if self.qemu_platform_runbook.is_host_remote():
+                self.host_node.tools[Iptables].start_forwarding(10022, address, 22)
```

Comment:
This doesn't seem like it'll support testing multiple VMs on the same host.

What I recommend you do is allow the user to set the libvirt network that the VM connects to. Then the user can setup their own network bridge on the host, create a libvirt "network" that points to that bridge, and then set tell LISA to use that "network" for the VMs. Then the test runner will be able to access the VMs directly without needing to deal with port forwarding.

Reply:
We can use the consistent way for local and remote hosts. It will simplify the code logic. The only difference is the localhost doesn't need connection info.

Reply:
In case of port forwarding, we do have special logic for remote hosts (set up iptable rules for forwarding). So consistent way is not possible.

Reply:
Anyway, I think it can make the logic simpler, if the port forwarding or other approaches are used on local too.
If to use port forwarding, one way is to set a start port range from high end like 30000, and map the ports one by one incrementally, if a port is unused.

Reply:
We probably shouldn't use port forward on local, as people can run these tests directly on their dev box and we don't want to be opening random inbound ports if it isn't necessary.

Reply:
I still think using libvirt networks is the correct way to solve this problem. Though I realize this might be a non-trivial change. So, I am happy if you delay this to a subsequent task.

Reply:
I need to figure out the bridged network thing. I had tried it at first but it didn't work. I am inclined to doing it in a follow up PR.

Just to understand, in what cases do we need multiple VMs on the same host? When I do implement it how do I test it?

Reply:
You can create a simple test function with the following annotation:
```python3
@TestCaseMetadata(
    description="",
    requirement=node_requirement(
        node=schema.NodeSpace(
            node_count=2,
        )
    ),
)
```

There are a bunch of "Microsoft suite" tests that use multiple nodes, I think mainly around testing networking. I am using it to test multi-node Kubernetes clusters.

Reply:
@cwize1 The field `schema.Environment.topology` is used to differentiate different network topology, but it's not used so far. The value is always subnet. Please let me know, if you need to test different topology, we can discuss the requirement, and think about how to support it.

Reply:
@squirrelsc Unfortunately, unlike with Hyper-V which has a fairly unified networking API, libvirt/QEMU doesn't provide much assistance for networking. Unless you are doing something super simple (e.g. using the default NAT network), then a developer has to manually create the network outside of libvirt and then point the libvirt VM at that network. Typically, either Linux kernel bridges or OVS (Open vSwitch) are used but both have their challenges. OVS is a big piece of software that is complicated to use. And every Linux distro has their own networking manager API that you have to us to provision networks, such as kernel bridges. I don't think it is worth trying to pull that complexity into LISA.

Reply:
Thank you for the thoughts. IMO, it depends on test scenarios, and doesn't need to fully automate all things. The lab environments is complex, and it's acceptable for preconfigured steps. If some servers are setup with some topologies, LISA just needs to use it, and keep it unchanged. The LISA needs to be aware of the topology, and assign test cases by its requirements. If the different topology tests is needed from a test case, the platform checks the settings (from runbook) of each server, and then allocate VMs from matched servers. The ADO agents supports it by "capabilities". The capabilities just a couple of key/value pairs, it's very similar like what I want to do.